File: examples/cpp/Misc/PipelineDebugging/get_pipeline_state.cpp
#include <iostream>
#include <opencv2/opencv.hpp>

#include "depthai/depthai.hpp"

int main() {
    dai::Pipeline pipeline;
    pipeline.enablePipelineDebugging();

    auto monoLeft = pipeline.create<dai::node::Camera>()->build(dai::CameraBoardSocket::CAM_B);
    auto monoRight = pipeline.create<dai::node::Camera>()->build(dai::CameraBoardSocket::CAM_C);

    auto stereo = pipeline.create<dai::node::StereoDepth>();

    auto monoLeftOut = monoLeft->requestFullResolutionOutput();
    auto monoRightOut = monoRight->requestFullResolutionOutput();

    monoLeftOut->link(stereo->left);
    monoRightOut->link(stereo->right);

    stereo->setRectification(true);
    stereo->setExtendedDisparity(true);
    stereo->setLeftRightCheck(true);

    auto depthQueue = stereo->depth.createOutputQueue();

    pipeline.start();
    while(pipeline.isRunning()) {
        auto depth = depthQueue->get<dai::ImgFrame>();
        cv::imshow("depth", dai::utility::colorizeDepthFrame(*depth).getCvFrame());

        int key = cv::waitKey(1);
        if(key == 'q') {
            break;
        } else if(key == 's') {
            auto state = pipeline.getPipelineState().nodes().detailed();
            try {
                // Assuming these APIs exist similarly in C++
                auto pipelineState = pipeline.getPipelineState().nodes().detailed();

                for(const auto& [nodeId, nodeState] : pipelineState.nodeStates) {
                    std::string nodeName = pipeline.getNode(nodeId)->getName();

                    std::cout << "\n# State for node " << nodeName << " (" << nodeId << "):\n";

                    std::cout << "## State: " << (int)nodeState.state << "\n";

                    std::cout << "## mainLoopTiming: " << (nodeState.mainLoopTiming.isValid() ? "" : "invalid") << "\n";
                    if(nodeState.mainLoopTiming.isValid()) {
                        std::cout << "-----\n" << nodeState.mainLoopTiming.str() << "\n-----\n";
                    }

                    std::cout << "## inputsGetTiming: " << (nodeState.inputsGetTiming.isValid() ? "" : "invalid") << "\n";
                    if(nodeState.inputsGetTiming.isValid()) {
                        std::cout << "-----\n" << nodeState.inputsGetTiming.str() << "\n-----\n";
                    }

                    std::cout << "## outputsSendTiming: " << (nodeState.outputsSendTiming.isValid() ? "" : "invalid") << "\n";
                    if(nodeState.outputsSendTiming.isValid()) {
                        std::cout << "-----\n" << nodeState.outputsSendTiming.str() << "\n-----\n";
                    }

                    std::cout << "## inputStates: " << (nodeState.inputStates.empty() ? "empty" : "") << "\n";
                    for(const auto& [inputName, inputState] : nodeState.inputStates) {
                        if(inputState.isValid())
                            std::cout << "### " << inputName << ":\n-----" << inputState.str() << "\n-----\n";
                        else
                            std::cout << "### " << inputName << ": invalid\n";
                    }

                    std::cout << "## outputStates: " << (nodeState.outputStates.empty() ? "empty" : "") << "\n";
                    for(const auto& [outputName, outputState] : nodeState.outputStates) {
                        std::cout << "### " << outputName << ":\n-----" << outputState.str() << "\n-----\n";
                    }

                    std::cout << "## otherTimings: " << (nodeState.otherTimings.empty() ? "empty" : "") << "\n";
                    for(const auto& [otherName, otherTiming] : nodeState.otherTimings) {
                        std::cout << "### " << otherName << ":\n-----" << otherTiming.str() << "\n-----\n";
                    }
                }
            } catch(const std::runtime_error& e) {
                std::cerr << "Error getting pipeline state: " << e.what() << "\n";
            }
        }
    }

    pipeline.stop();

    return 0;
}

File: examples/python/Misc/PipelineDebugging/get_pipeline_state.py
import depthai as dai
import numpy as np
import cv2

with dai.Pipeline() as pipeline:
    # Pipeline debugging is disabled by default.
    # You can also enable it by setting the DEPTHAI_PIPELINE_DEBUGGING environment variable to '1'
    pipeline.enablePipelineDebugging(True)

    monoLeft = pipeline.create(dai.node.Camera).build(dai.CameraBoardSocket.CAM_B)
    monoRight = pipeline.create(dai.node.Camera).build(dai.CameraBoardSocket.CAM_C)
    stereo = pipeline.create(dai.node.StereoDepth)

    # Linking
    monoLeftOut = monoLeft.requestFullResolutionOutput()
    monoRightOut = monoRight.requestFullResolutionOutput()
    monoLeftOut.link(stereo.left)
    monoRightOut.link(stereo.right)

    stereo.setRectification(True)
    stereo.setExtendedDisparity(True)
    stereo.setLeftRightCheck(True)

    disparityQueue = stereo.disparity.createOutputQueue()

    colorMap = cv2.applyColorMap(np.arange(256, dtype=np.uint8), cv2.COLORMAP_JET)
    colorMap[0] = [0, 0, 0]  # to make zero-disparity pixels black

    pipeline.start()
    maxDisparity = 1
    while pipeline.isRunning():
        disparity = disparityQueue.get()
        assert isinstance(disparity, dai.ImgFrame)
        npDisparity = disparity.getFrame()
        maxDisparity = max(maxDisparity, np.max(npDisparity))
        colorizedDisparity = cv2.applyColorMap(((npDisparity / maxDisparity) * 255).astype(np.uint8), colorMap)
        cv2.imshow("disparity", colorizedDisparity)
        key = cv2.waitKey(1)
        if key == ord('q'):
            pipeline.stop()
            break
        elif key == ord('s'):
            try:
                # If pipeline debugging is disabled, this will raise an exception
                pipelineState = pipeline.getPipelineState().nodes().detailed()
                for nodeId, nodeState in pipelineState.nodeStates.items():
                    nodeName = pipeline.getNode(nodeId).getName()
                    print(f"\n# State for node {pipeline.getNode(nodeId).getName()} ({nodeId}):")
                    print(f"## State: {nodeState.state}")
                    print(f"## mainLoopTiming: {'invalid' if not nodeState.mainLoopTiming.isValid() else ''}")
                    if(nodeState.mainLoopTiming.isValid()):
                        print("-----")
                        print(nodeState.mainLoopTiming)
                        print("-----")
                    print(f"## inputsGetTiming: {'invalid' if not nodeState.inputsGetTiming.isValid() else ''}")
                    if(nodeState.inputsGetTiming.isValid()):
                        print("-----")
                        print(nodeState.inputsGetTiming)
                        print("-----")
                    print(f"## outputsSendTiming: {'invalid' if not nodeState.outputsSendTiming.isValid() else ''}")
                    if(nodeState.outputsSendTiming.isValid()):
                        print("-----")
                        print(nodeState.outputsSendTiming)
                        print("-----")
                    print(f"## inputStates: {'empty' if not nodeState.inputStates else ''}")
                    for inputName, inputState in nodeState.inputStates.items():
                        if inputState.isValid():
                            print(f"### {inputName}:\n-----{inputState}\n-----")
                        else:
                            print(f"### {inputName}: invalid")
                    print(f"## outputStates: {'empty' if not nodeState.outputStates else ''}")
                    for outputName, outputState in nodeState.outputStates.items():
                        print(f"### {outputName}:\n-----{outputState}\n-----")
                    print(f"## otherTimings: {'empty' if not nodeState.otherTimings else ''}")
                    for otherName, otherTiming in nodeState.otherTimings.items():
                        print(f"### {otherName}:\n-----{otherTiming}\n-----")
            except Exception as e:
                print("Error getting pipeline state:", e)
